plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,722 @@
+      for (const auto& slice : table_slices) {
+        auto final_slice = filter(slice, expr, ids);
+        if (final_slice) {
+          num_hits += final_slice->rows();
+          self->send(extract.sink, *final_slice);
+        }
+      }
+      return num_hits;
+    },
+  };
+  return caf::visit(handle_query, query.cmd);
+}
+
+std::shared_ptr<arrow::Schema> parse_arrow_schema_from_metadata(
+  const std::shared_ptr<parquet::FileMetaData>& parquet_metadata) {
+  if (!parquet_metadata)
+    return {};
+  auto arrow_metadata
+    = parquet_metadata->key_value_metadata()->Get("ARROW:schema");
+  if (!arrow_metadata.ok())
+    return {};
+  auto decoded = detail::base64::decode(*arrow_metadata);
+  auto schema_buf = arrow::Buffer(decoded);
+  arrow::ipc::DictionaryMemo dict_memo;
+  arrow::io::BufferReader input(schema_buf);
+  auto arrow_schema = arrow::ipc::ReadSchema(&input, &dict_memo);
+  if (!arrow_schema.ok())
+    return {};
+  return *arrow_schema;
+}
+
+caf::expected<std::shared_ptr<arrow::Table>>
+read_parquet_buffer(const chunk_ptr& chunk) {
+  VAST_ASSERT(chunk);
+  auto bufr = std::make_shared<arrow::io::BufferReader>(as_arrow_buffer(chunk));
+  auto parquet_reader = parquet::ParquetFileReader::Open(bufr);
+  auto arrow_schema
+    = parse_arrow_schema_from_metadata(parquet_reader->metadata());
+  std::unique_ptr<parquet::arrow::FileReader> file_reader{};
+  if (!parquet::arrow::OpenFile(bufr, arrow::default_memory_pool(),
+                                &file_reader)
+         .ok())
+    return caf::exit_reason::unhandled_exception;
+  std::shared_ptr<arrow::Table> table{};
+  if (!file_reader->ReadTable(&table).ok())
+    return caf::exit_reason::unhandled_exception;
+  return align_table_to_schema(arrow_schema, table);
+}
+
+auto init_parquet(caf::unit_t&) {
+  // doing nothing: as we're writing the file in one pass at the end into a
+  // memory buffer, there's no file opening going on here.
+}
+
+auto add_table_slices(
+  system::store_builder_actor::stateful_pointer<store_builder_state> self) {
+  return [self](caf::unit_t&, std::vector<table_slice>& batch) {
+    for (auto& slice : batch) {
+      if (self->state.vast_type_) {
+        VAST_ASSERT(*self->state.vast_type_ == slice.layout());
+      } else {
+        self->state.vast_type_ = slice.layout();
+      }
+      self->state.num_rows_ += slice.rows();
+      self->state.table_slices_.push_back(slice);
+    }
+    VAST_TRACE("{} received batch of {} table slices for partition {}", *self,
+               batch.size(), self->state.id_);
+  };
+}
+
+std::shared_ptr<parquet::WriterProperties>
+writer_properties(const configuration& config) {
+  auto builder = parquet::WriterProperties::Builder{};
+  builder.created_by("VAST telemetry engine")
+    ->max_row_group_length(detail::narrow_cast<int64_t>(config.row_group_size))
+    ->enable_dictionary()
+    ->compression(parquet::Compression::ZSTD)
+    ->compression_level(9)
```

Comment:
Why 9 instead of the default level? Should we make this configurable instead?

Reply:
I used this [blog post](https://thedatafreak.medium.com/parquet-zstd-vs-gzip-740a83571ecd) as a baseline, but I'm adding a config option and just default to 9 instead.